### DISCUSSION TITLE
Feature Request - Improve typings when passing initialData

### DISCUSSION THREAD

**Discussion by iway1** [TanStack/query#5080] [Ideas] [answered]
Any chance we could get the type of `data` to not be undefined in cases where we pass `initialData`? we're more or less guaranteed the cache will have data in these cases right and it'd make it easier to code with if we could have type of `.data` just be whatever type is returned from the query function I think.

This would make it a lot easier to handle the typings when doing SSR specifically:

```tsx
const Page = ({posts})=>{
  const query = useQuery({
    initialData: posts
  }) 
  query.data // this is typeof posts | undefined, why not just typeof posts?
}
```

**Comment by TkDodo** [comment, marked as answer]
that was contributed some time ago - we have tests for this:

https://github.com/TanStack/query/blob/9be63f4c13cda2f8cfd9cd63c76f23fce3b06e5c/packages/react-query/src/__tests__/useQuery.types.test.tsx#L7-L24
